addons/l10n_id_efaktur/models/account_move.py

```diff
@@ -0,0 +1,314 @@
+    l10n_id_upload_user = fields.Many2one('res.users', 'Upload User', readonly=True, copy=False)
+    l10n_id_attachment_id = fields.Many2one('ir.attachment', readonly=True, copy=False)
+    l10n_id_date_validated = fields.Datetime('Upload Date', readonly=True, copy=False)
+    l10n_id_csv_created = fields.Boolean('CSV Created', compute='_compute_csv_created', copy=False)
+    l10n_id_kode_transaksi = fields.Selection([
+            ('01', '01 Kepada Pihak yang Bukan Pemungut PPN (Customer Biasa)'),
+            ('02', '02 Kepada Pemungut Bendaharawan (Dinas Kepemerintahan)'),
+            ('03', '03 Kepada Pemungut Selain Bendaharawan (BUMN)'),
+            ('04', '04 DPP Nilai Lain (PPN 1%)'),
+            ('06', '06 Penyerahan Lainnya (Turis Asing)'),
+            ('07', '07 Penyerahan yang PPN-nya Tidak Dipungut (Kawasan Ekonomi Khusus/ Batam)'),
+            ('08', '08 Penyerahan yang PPN-nya Dibebaskan (Impor Barang Tertentu)'),
+            ('09', '09 Penyerahan Aktiva ( Pasal 16D UU PPN )'),
+        ], string='Kode Transaksi', help='Dua digit pertama nomor pajak',
+        readonly=True, states={'draft': [('readonly', False)]}, copy=False)
+    l10n_id_need_kode_transaksi = fields.Boolean(compute='_compute_need_kode_transaksi')
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        self.l10n_id_kode_transaksi = self.partner_id.l10n_id_kode_transaksi
+        return super(AccountMove, self)._onchange_partner_id()
+
+    def _inverse_tax_number(self):
+        for record in self:
+            if record.type == 'in_invoice':
+                if record.l10n_id_efaktur_id:
+                    record.l10n_id_efaktur_id.name = record.l10n_id_tax_number
+                elif record.l10n_id_tax_number:
+                    self.env['l10n_id_efaktur.efaktur'].create({
+                        'name': record.l10n_id_tax_number,
+                        'invoice_id': [(6, 0, record.ids)],
+                        'company_id': record.company_id.id,
+                    })
+            elif record.l10n_id_efaktur_id.name != record.l10n_id_tax_number:
+                raise UserError(_("You can't change the number manually for a Customer Invoice"))
+
+    @api.depends('l10n_id_efaktur_id')
+    def _compute_tax_number(self):
+        for record in self:
+            record.l10n_id_tax_number = record.l10n_id_efaktur_id.name
+
+    @api.depends('l10n_id_attachment_id')
+    def _compute_csv_created(self):
+        for record in self:
+            record.l10n_id_csv_created = bool(record.l10n_id_attachment_id)
+
+    @api.depends('partner_id', 'line_ids.tag_ids')
+    def _compute_need_kode_transaksi(self):
+        for move in self:
+            move.l10n_id_need_kode_transaksi = move.partner_id.l10n_id_pkp and not move.l10n_id_efaktur_id and move.type == 'out_invoice'
+
+    @api.constrains('l10n_id_kode_transaksi', 'line_ids')
+    def _constraint_kode_ppn(self):
```

Comment:
Why bother with constraints and not just do these things at validation?

Reply:
(Please highlight the whole function when you do the review so that I don't have to find it in the diff to know what you are talking about)

Why do it at validation and not with a constraint? I find it cleaner and safer.

Reply:
If you want to ensure that constraints are respected, you need to do it every time something is changed, not only once and let the user change it afterwards.